BootloaderCorePkg/Stage1B/Stage1B.c

```diff
@@ -7,26 +7,52 @@
 
 #include "Stage1B.h"
 
+
+
+
 /**
   Function to extend stage2 hash
 
   @param[in]  CbInfo    Component Call Back Info
 
 **/
 VOID
-ExtendStage (
+ExtendStageHash (
   IN  COMPONENT_CALLBACK_INFO   *CbInfo
   )
 {
-  UINT8                     BootMode;
+  UINT8                BootMode;
+  UINT8                DigestHash[HASH_DIGEST_MAX];
+  HASH_ALG_TYPE        MbHashType;
+  TPMI_ALG_HASH        MbTmpAlgHash;
+  RETURN_STATUS        Status;
+
+  //Convert Measured boot Hash Mask to HASH_ALG_TYPE (CryptoLib)
+  MbHashType   = GetCryptoHashAlg(PcdGet32(PcdMeasuredBootHashMask));
+
+  //Convert Measured boot Hash Mask to TPMI_ALG_HASH (TPM ALG ID)
+  MbTmpAlgHash = (TPMI_ALG_HASH) GetTpmHashAlg(PcdGet32(PcdMeasuredBootHashMask));
 
   //Check the boot mode
   BootMode = GetBootMode();
   if (MEASURED_BOOT_ENABLED() && (BootMode != BOOT_ON_S3_RESUME)) {
     //Extend  hash if ComponentType is stage2
     if ((CbInfo != NULL ) && (CbInfo->ComponentType == COMP_TYPE_STAGE_2)) {
-      TpmExtendStageHash((UINT8) CbInfo->ComponentType, CbInfo->HashData, CbInfo->HashAlg,
-                          CbInfo->CompBuf, CbInfo->CompLen);
+      // Check Hash data alg match to PcdMeasuredBootHashMask
+      if ((CbInfo->HashAlg == MbHashType) && (CbInfo->HashData != NULL)) {
+        // Extend CbInfo->HashData if hashalg is valid
+        TpmExtendPcrAndLogEvent (0, MbTmpAlgHash, CbInfo->HashData,
+                                EV_POST_CODE, POST_CODE_STR_LEN, (UINT8 *)EV_POSTCODE_INFO_POST_CODE);
+
+      } else {
+        // Get Hash to extend based on component type and component src addresss
+        Status = GetHashToExtend ((UINT8) CbInfo->ComponentType,
+                                    MbHashType, CbInfo->CompBuf, CbInfo->CompLen, DigestHash);
+        if (Status == EFI_SUCCESS){
+          TpmExtendPcrAndLogEvent (0, MbTmpAlgHash, DigestHash,
+                                EV_POST_CODE, POST_CODE_STR_LEN, (UINT8 *)EV_POSTCODE_INFO_POST_CODE);
+        }
+      }

```

Comment:
What about this?
UINT8 *HashPtr;

if ((CbInfo->HashAlg == MbHashType) && (CbInfo->HashData != NULL)) {
  HashPtr = CbInfo->HashData;
  Status = EFI_SUCCESS;
} else {
  // Get Hash to extend based on component type and component src addresss
  Status = GetHashToExtend ((UINT8) CbInfo->ComponentType, MbHashType, CbInfo->CompBuf, CbInfo->CompLen, DigestHash);								
  HashPtr = DigestHash;
}

if ((Status == EFI_SUCCESS) && (HashPtr != NULL)) {
  TpmExtendPcrAndLogEvent (0, MbTmpAlgHash, HashPtr, EV_POST_CODE, POST_CODE_STR_LEN, (UINT8 *)EV_POSTCODE_INFO_POST_CODE);
}

Reply:
yes. We could do. Thanks!